File: platform/tests/seed_from_excel.mjs
import { createClient } from "@supabase/supabase-js";
import dotenv from "dotenv";
import { resolve, dirname } from "path";
import { fileURLToPath } from "url";
import XLSX from "xlsx";
import fs from "fs";

const __dirname = dirname(fileURLToPath(import.meta.url));
dotenv.config({ path: resolve(__dirname, "../../.env") });

const sb = createClient(process.env.SUPABASE_URL, process.env.SUPABASE_ANON_KEY);
const i = sb.schema("internal");

// ── Read Excel ──
const xlsxPath = resolve(__dirname, "../../docs/店铺信息.xlsx");
const wb = XLSX.readFile(xlsxPath);
const ws = wb.Sheets[wb.SheetNames[0]];
const rows = XLSX.utils.sheet_to_json(ws, { header: 1 });

// Column indices: 0=title, 2=operators, 12=shop_id, 18=shop_name
const excelRows = [];
for (let r = 1; r < rows.length; r++) {
  const row = rows[r];
  const title = String(row[0] || "");
  const operatorsStr = String(row[2] || "");
  const shopIdRaw = String(row[12] || "");
  const shopName = String(row[18] || "");
  const operators = operatorsStr.split(",").map(s => s.trim()).filter(Boolean);
  if (!operators.length) continue;
  excelRows.push({ title, operators, shopId: shopIdRaw, shopName });
}

// ── Get DB shops ──
const { data: dbShops } = await sb.from("shops").select("shop_id, name, platform");
console.log(`Excel: ${excelRows.length} 行, DB: ${dbShops.length} 店铺`);

// ── Match by numeric ID ──
function extractNums(s) {
  return (s.match(/\d{7,}/g) || []);
}

const matches = [];
const unmatched = [];
for (const row of excelRows) {
  const nums = [...extractNums(row.title), ...extractNums(row.shopId), ...extractNums(row.shopName)];
  let found = null;
  for (const num of nums) {
    found = dbShops.find(s => s.name.includes(num));
    if (found) break;
  }
  if (found) matches.push({ ...row, dbShop: found });
  else unmatched.push(row);
}

console.log(`匹配: ${matches.length}, 未匹配: ${unmatched.length}`);
for (const r of unmatched) console.log("  未匹配:", r.title.slice(0, 60), "|", r.shopId);

// ── Operators ──
const allOps = new Set();
for (const m of matches) for (const o of m.operators) allOps.add(o);
console.log(`\n运营者 (${allOps.size}人):`, [...allOps].join(", "));

// ── Clean & Seed ──
console.log("\n清理旧数据...");
await i.from("shop_operators").delete().neq("id", 0);
await i.from("operators").delete().neq("id", 0);
await i.from("operator_groups").delete().neq("id", 0);

// Groups
const groups = ["步凡者（巴西）", "美度邦（巴西）", "步凡者（台湾）"];
const groupIds = {};
for (const name of groups) {
  const { data } = await i.from("operator_groups").insert({ name }).select().single();
  if (data) { groupIds[name] = data.id; console.log("分组:", name, data.id); }
}

// Operators
const opIds = {};
for (const name of allOps) {
  const { data } = await i.from("operators").insert({ name }).select().single();
  if (data) { opIds[name] = data.id; console.log("人员:", name, data.id); }
}

// Bindings: each shop→operator pair from Excel
let bound = 0, errors = 0;
for (const m of matches) {
  const sid = m.dbShop.shop_id;
  const gid = groupIds["步凡者（巴西）"];
  for (const opName of m.operators) {
    const oid = opIds[opName];
    if (!oid) continue;
    const { error } = await i.from("shop_operators").insert({
      shop_id: sid, operator_id: oid, group_id: gid,
    });
    if (error) {
      if (!error.message.includes("duplicate")) {
        errors++;
        if (errors <= 3) console.error("绑定失败:", m.dbShop.name, "→", opName, error.message);
      }
    } else {
      bound++;
    }
  }
}

console.log(`\n完成: ${bound} 条绑定, ${errors} 个错误`);

File: platform/tests/seed_from_excel.py
"""从 docs/店铺信息.xlsx 读取店铺-运营者绑定，seed 到 Supabase"""
import os, re, sys
import openpyxl
from dotenv import load_dotenv
from supabase import create_client

load_dotenv(os.path.join(os.path.dirname(__file__), "../../.env"))

sb = create_client(os.environ["SUPABASE_URL"], os.environ["SUPABASE_ANON_KEY"])

# ── 读取 Excel ──
xlsx = openpyxl.load_workbook(os.path.join(os.path.dirname(__file__), "../../docs/店铺信息.xlsx"))
ws = xlsx["DefaultSheet"]

excel_rows = []
for row in ws.iter_rows(min_row=2, values_only=True):
    title = str(row[0] or "")
    operators_str = str(row[2] or "")
    shop_id_raw = str(row[12] or "")
    shop_name = str(row[18] or "")
    operators = [o.strip() for o in operators_str.split(",") if o.strip()]
    if not operators:
        continue
    excel_rows.append({"title": title, "operators": operators, "shop_id": shop_id_raw, "shop_name": shop_name})

# ── 获取 DB 店铺 ──
db_shops = sb.table("shops").select("shop_id, name, platform").execute().data
print(f"Excel行数: {len(excel_rows)}, DB店铺数: {len(db_shops)}")

# ── 匹配 ──
def extract_nums(s):
    return re.findall(r"\d{7,}", s)

matches = []
unmatched = []
for row in excel_rows:
    nums = extract_nums(row["title"]) + extract_nums(row["shop_id"]) + extract_nums(row["shop_name"])
    found = None
    for num in nums:
        found = next((s for s in db_shops if num in s["name"]), None)
        if found:
            break
    if found:
        matches.append({**row, "db_shop": found})
    else:
        unmatched.append(row)

print(f"匹配成功: {len(matches)}, 未匹配: {len(unmatched)}")
for r in unmatched:
    print(f"  未匹配: {r['title'][:60]} | shop_id={r['shop_id']}")

# ── 运营者汇总 ──
all_ops = set()
for m in matches:
    for o in m["operators"]:
        all_ops.add(o)
print(f"\n运营者 ({len(all_ops)}人): {', '.join(sorted(all_ops))}")

# ── 清理并 Seed ──
schema = sb.schema("internal")
print("\n清理旧数据...")
schema.table("shop_operators").delete().neq("id", 0).execute()
schema.table("operators").delete().neq("id", 0).execute()
schema.table("operator_groups").delete().neq("id", 0).execute()

# 分组
groups = ["步凡者（巴西）", "美度邦（巴西）", "步凡者（台湾）"]
group_ids = {}
for name in groups:
    r = schema.table("operator_groups").insert({"name": name}).execute()
    if r.data:
        group_ids[name] = r.data[0]["id"]
        print(f"分组: {name} id={r.data[0]['id']}")

# 人员
op_ids = {}
for name in all_ops:
    r = schema.table("operators").insert({"name": name}).execute()
    if r.data:
        op_ids[name] = r.data[0]["id"]
        print(f"人员: {name} id={r.data[0]['id']}")

# 绑定：一个店铺可能对应多个运营者（Excel中逗号分隔）
bound = 0
errors = 0
for m in matches:
    sid = m["db_shop"]["shop_id"]
    gid = group_ids.get("步凡者（巴西）")  # 默认分组
    for op_name in m["operators"]:
        oid = op_ids.get(op_name)
        if not oid:
            continue
        try:
            r = schema.table("shop_operators").insert({
                "shop_id": sid, "operator_id": oid, "group_id": gid
            }).execute()
            if r.data:
                bound += 1
        except Exception as e:
            errors += 1
            if errors <= 3:
                print(f"绑定失败: {m['db_shop']['name']} → {op_name}: {e}")

print(f"\n完成: {bound} 条绑定, {errors} 个错误")
